pkg/log/syslog/syslog.go

```diff
@@ -0,0 +1,248 @@
+// Copyright 2016-2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package syslog
+
+import (
+	"io"
+	"net"
+
+	"github.com/RackSec/srslog"
+	"github.com/Sirupsen/logrus"
+)
+
+type SyslogConfig struct {
+	Network   string
+	RAddr     string
+	Tag       string
+	Priority  srslog.Priority
+	Formatter Formatter
+}
+
+type dialer interface {
+	dial(*SyslogConfig) (Writer, error)
+}
+
+type Hook struct {
+	entries chan *logrus.Entry
+	writer  Writer
+	cfg     SyslogConfig
+
+	// for testing
+	running chan struct{}
+}
+
+type Writer interface {
+	io.WriteCloser
+
+	Emerg(string) error
+	Crit(string) error
+	Err(string) error
+	Warning(string) error
+	Info(string) error
+	Debug(string) error
+}
+
+const maxLogBuffer = 100
+
+type Formatter int
+
+const (
+	RFC3164 Formatter = iota
+)
+
+func NewHook(cfg *SyslogConfig, d dialer) (*Hook, error) {
+	hook := &Hook{
+		entries: make(chan *logrus.Entry, maxLogBuffer),
+		running: make(chan struct{}),
+	}
+
+	if d == nil {
+		d = &defaultDialer{
+			d: &syslogDialer{},
+		}
+	}
+
+	var err error
+	hook.writer, err = d.dial(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return hook, nil
+}
+
+type defaultDialer struct {
+	// for testing purposes
+	d dialer
+}
+
+func (d *defaultDialer) dial(cfg *SyslogConfig) (Writer, error) {
+	w, err := d.d.dial(cfg)
+	if err != nil {
+		switch err.(type) {
+		case *net.AddrError, *net.ParseError:
+			return nil, err
```

Comment:
We can't put out a log message as that would cause an infinite loop since the log call is going to end up in this code, and this code would call log again ...